File: integration-tests/csharp/Tests/ConstantsTests.cs
// Copyright 2026 KONGSBERG
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice,
//    this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice,
//    this list of conditions and the following disclaimer in the documentation
//    and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors
//    may be used to endorse or promote products derived from this software
//    without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

using Xunit;
using ConstantTypes;
using LargeIntegerTypes;

namespace IntegrationTests;

public class ConstantsTests
{
    [Fact]
    public void IntConst_HasCorrectValue()
    {
        Assert.Equal(42, ConstantTypes.Constants.IntConst);
    }

    [Fact]
    public void UintConst_HasCorrectValue()
    {
        Assert.Equal(100u, ConstantTypes.Constants.UintConst);
    }

    [Fact]
    public void ShortConst_HasCorrectValue()
    {
        Assert.Equal(-10, ConstantTypes.Constants.ShortConst);
    }

    [Fact]
    public void LonglongConst_HasCorrectValue()
    {
        Assert.Equal(9999999999L, ConstantTypes.Constants.LonglongConst);
    }

    [Fact]
    public void DoubleConst_HasCorrectValue()
    {
        Assert.Equal(3.14159, ConstantTypes.Constants.DoubleConst, 5);
    }

    [Fact]
    public void FloatConst_HasCorrectValue()
    {
        Assert.Equal(2.5f, ConstantTypes.Constants.FloatConst);
    }

    [Fact]
    public void StringConst_HasCorrectValue()
    {
        Assert.Equal("hello world", ConstantTypes.Constants.StringConst);
    }

    [Fact]
    public void BoolTrue_HasCorrectValue()
    {
        Assert.True(ConstantTypes.Constants.BoolTrue);
    }

    [Fact]
    public void BoolFalse_HasCorrectValue()
    {
        Assert.False(ConstantTypes.Constants.BoolFalse);
    }

    [Fact]
    public void OctetConst_HasCorrectValue()
    {
        Assert.Equal(255, ConstantTypes.Constants.OctetConst);
    }

    [Fact]
    public void ChainedConstants_HaveCorrectValues()
    {
        Assert.Equal(10, ConstantTypes.Constants.Chain1);
        Assert.Equal(ConstantTypes.Constants.Chain1, ConstantTypes.Constants.Chain2);
        Assert.Equal(ConstantTypes.Constants.Chain2, ConstantTypes.Constants.Chain3);
        Assert.Equal(ConstantTypes.Constants.Chain3, ConstantTypes.Constants.Chain4);
        Assert.Equal(ConstantTypes.Constants.Chain4, ConstantTypes.Constants.Chain5);
    }

    [Fact]
    public void ArithmeticConstants_HaveCorrectValues()
    {
        Assert.Equal(100, ConstantTypes.Constants.ArithBase);
        Assert.Equal(200, ConstantTypes.Constants.ArithDoubled);
        Assert.Equal(400, ConstantTypes.Constants.ArithQuadrupled);
        Assert.Equal(800, ConstantTypes.Constants.ArithOctupled);
    }

    [Fact]
    public void MathConstants_HaveCorrectValues()
    {
        Assert.Equal(5, ConstantTypes.Constants.Math1);
        Assert.Equal(15, ConstantTypes.Constants.Math2);
        Assert.Equal(30, ConstantTypes.Constants.Math3);
        Assert.Equal(25, ConstantTypes.Constants.Math4);
        Assert.Equal(5, ConstantTypes.Constants.Math5);
    }

    [Fact]
    public void NegationConstants_HaveCorrectValues()
    {
        Assert.Equal(-50, ConstantTypes.Constants.Negative);
        Assert.Equal(50, ConstantTypes.Constants.Negated);
        Assert.Equal(-50, ConstantTypes.Constants.DoubleNegated);
    }

    [Fact]
    public void BitwiseConstants_HaveCorrectValues()
    {
        Assert.Equal(15, ConstantTypes.Constants.BitsA);
        Assert.Equal(240, ConstantTypes.Constants.BitsB);
        Assert.Equal(255, ConstantTypes.Constants.BitsOr);
        Assert.Equal(255, ConstantTypes.Constants.BitsAnd);
        Assert.Equal(240, ConstantTypes.Constants.BitsXor);
        Assert.Equal(16, ConstantTypes.Constants.BitsShiftLeft);
        Assert.Equal(16, ConstantTypes.Constants.BitsShiftRight);
    }

    [Fact]
    public void FloatConstants_HaveCorrectValues()
    {
        Assert.Equal(1.0, ConstantTypes.Constants.FloatA);
        Assert.Equal(1.5, ConstantTypes.Constants.FloatB);
        Assert.Equal(3.0, ConstantTypes.Constants.FloatC);
        Assert.Equal(0.75, ConstantTypes.Constants.FloatD);
    }

    [Fact]
    public void PriorityValue_HasCorrectValue()
    {
        Assert.Equal(Priority.High, ConstantTypes.Constants.PriorityValue);
        Assert.Equal(100, (int)ConstantTypes.Constants.PriorityValue);
    }

    [Fact]
    public void StringConstants_HaveCorrectValues()
    {
        Assert.Equal("Hello", ConstantTypes.Constants.Greeting);
        Assert.Equal("Goodbye", ConstantTypes.Constants.Farewell);
    }

    [Fact]
    public void ParenthesizedConstants_HaveCorrectValues()
    {
        Assert.Equal(30, ConstantTypes.Constants.ParenA);
        Assert.Equal(20, ConstantTypes.Constants.ParenB);
        Assert.Equal(25, ConstantTypes.Constants.ParenC);
    }

    [Fact]
    public void ModuloConstants_HaveCorrectValues()
    {
        Assert.Equal(2, ConstantTypes.Constants.ModA);
        Assert.Equal(2, ConstantTypes.Constants.ModB);
    }

    // Large integer tests (LargeIntegerTypes namespace)
    [Fact]
    public void LargeIntegers_OctetBounds()
    {
        Assert.Equal(255, LargeIntegerTypes.Constants.OctetMax);
        Assert.Equal(0, LargeIntegerTypes.Constants.OctetMin);
    }

    [Fact]
    public void LargeIntegers_ShortBounds()
    {
        Assert.Equal(32767, LargeIntegerTypes.Constants.ShortMax);
        Assert.Equal(-32768, LargeIntegerTypes.Constants.ShortMin);
    }

    [Fact]
    public void LargeIntegers_UshortBounds()
    {
        Assert.Equal(65535, LargeIntegerTypes.Constants.UshortMax);
        Assert.Equal(0, LargeIntegerTypes.Constants.UshortMin);
    }

    [Fact]
    public void LargeIntegers_LongBounds()
    {
        Assert.Equal(2147483647, LargeIntegerTypes.Constants.LongMax);
        Assert.Equal(-2147483648, LargeIntegerTypes.Constants.LongMin);
    }

    [Fact]
    public void LargeIntegers_UlongBounds()
    {
        Assert.Equal(4294967295u, LargeIntegerTypes.Constants.UlongMax);
        Assert.Equal(0u, LargeIntegerTypes.Constants.UlongMin);
    }

    [Fact]
    public void LargeIntegers_LonglongBounds()
    {
        Assert.Equal(9223372036854775807L, LargeIntegerTypes.Constants.LonglongMax);
        Assert.Equal(-9223372036854775808L, LargeIntegerTypes.Constants.LonglongMin);
    }

    [Fact]
    public void LargeIntegers_UlonglongBounds()
    {
        Assert.Equal(18446744073709551615UL, LargeIntegerTypes.Constants.UlonglongMax);
        Assert.Equal(0UL, LargeIntegerTypes.Constants.UlonglongMin);
    }

    [Fact]
    public void LargeIntegers_HexValues()
    {
        Assert.Equal(0xDEADBEEFu, LargeIntegerTypes.Constants.HexDeadbeef);
        Assert.Equal(0xFFFFFFFFu, LargeIntegerTypes.Constants.HexFfffffff);
        Assert.Equal(0x123456789ABCDEF0L, LargeIntegerTypes.Constants.Hex64Bit);
    }

    [Fact]
    public void LargeIntegers_OctalValues()
    {
        Assert.Equal(511, LargeIntegerTypes.Constants.Octal777);
        Assert.Equal(2147483647, LargeIntegerTypes.Constants.OctalLarge);
    }

    [Fact]
    public void LargeIntegers_BoundaryArithmetic()
    {
        Assert.Equal(2147483646, LargeIntegerTypes.Constants.LongMaxMinusOne);
        Assert.Equal(9223372036854775806L, LargeIntegerTypes.Constants.LonglongMaxMinusOne);
    }

    [Fact]
    public void LargeIntFields_CanHoldLargeValues()
    {
        var fields = new LargeIntFields(
            BigSigned: long.MaxValue,
            BigUnsigned: ulong.MaxValue
        );
        Assert.Equal(long.MaxValue, fields.BigSigned);
        Assert.Equal(ulong.MaxValue, fields.BigUnsigned);
    }
}

File: integration-tests/java/src/test/java/integrationtests/ConstantsTests.java
// Copyright 2026 KONGSBERG
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice,
//    this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice,
//    this list of conditions and the following disclaimer in the documentation
//    and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors
//    may be used to endorse or promote products derived from this software
//    without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

package integrationtests;

import static org.junit.jupiter.api.Assertions.*;

import constant_types.*;
import large_integer_types.*;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

class ConstantsTests {

    @Test
    void int_constant() {
        assertEquals(42, INT_CONST.value);
    }

    @Test
    void uint_constant() {
        assertEquals(100, UINT_CONST.value);
    }

    @Test
    void short_constant() {
        assertEquals(-10, SHORT_CONST.value);
    }

    @Test
    void longlong_constant() {
        assertEquals(9999999999L, LONGLONG_CONST.value);
    }

    @Test
    void double_constant() {
        assertEquals(3.14159, DOUBLE_CONST.value, 0.00001);
    }

    @Test
    void float_constant() {
        assertEquals(2.5f, FLOAT_CONST.value, 0.001f);
    }

    @Test
    void string_constant() {
        assertEquals("hello world", STRING_CONST.value);
    }

    @Test
    void bool_constants() {
        assertTrue(BOOL_TRUE.value);
        assertFalse(BOOL_FALSE.value);
    }

    @Test
    void octet_constant() {
        assertEquals(255, OCTET_CONST.value);
    }

    @Test
    void constant_chain() {
        assertEquals(10, CHAIN_1.value);
        assertEquals(10, CHAIN_2.value);
        assertEquals(10, CHAIN_3.value);
        assertEquals(10, CHAIN_4.value);
        assertEquals(10, CHAIN_5.value);
    }

    @Test
    void arithmetic_chain() {
        assertEquals(100, ARITH_BASE.value);
        assertEquals(200, ARITH_DOUBLED.value);
        assertEquals(400, ARITH_QUADRUPLED.value);
        assertEquals(800, ARITH_OCTUPLED.value);
    }

    @Test
    void mixed_arithmetic_chain() {
        assertEquals(5, MATH_1.value);
        assertEquals(15, MATH_2.value);
        assertEquals(30, MATH_3.value);
        assertEquals(25, MATH_4.value);
        assertEquals(5, MATH_5.value);
    }

    @Test
    void negation() {
        assertEquals(-50, NEGATIVE.value);
        assertEquals(50, NEGATED.value);
        assertEquals(-50, DOUBLE_NEGATED.value);
    }

    @Test
    void bitwise_operations() {
        assertEquals(0x0F, BITS_A.value);
        assertEquals(0xF0, BITS_B.value);
        assertEquals(0xFF, BITS_OR.value);
        assertEquals(255, BITS_AND.value);
        assertEquals(240, BITS_XOR.value);
        assertEquals(16, BITS_SHIFT_LEFT.value);
        assertEquals(16, BITS_SHIFT_RIGHT.value);
    }

    @Test
    void float_chain() {
        assertEquals(1.0, FLOAT_A.value, 0.001);
        assertEquals(1.5, FLOAT_B.value, 0.001);
        assertEquals(3.0, FLOAT_C.value, 0.001);
        assertEquals(0.75, FLOAT_D.value, 0.001);
    }

    @Test
    void enum_constant_reference() {
        assertEquals(Priority.HIGH, PRIORITY_VALUE.value);
        assertEquals(Priority.HIGH, PRIORITY_CHAIN.value);
    }

    @Test
    void parenthesized_expressions() {
        assertEquals(30, PAREN_A.value);
        assertEquals(20, PAREN_B.value);
        assertEquals(25, PAREN_C.value);
    }

    @Test
    void modulo_operations() {
        assertEquals(2, MOD_A.value);
        assertEquals(2, MOD_B.value);
    }

    @Test
    void octet_limits() {
        assertEquals(255, OCTET_MAX.value);
        assertEquals(0, OCTET_MIN.value);
    }

    @Test
    void short_limits() {
        assertEquals(32767, SHORT_MAX.value);
        assertEquals(-32768, SHORT_MIN.value);
    }

    @Test
    void ushort_limits() {
        assertEquals(65535, USHORT_MAX.value);
        assertEquals(0, USHORT_MIN.value);
    }

    @Test
    void long_limits() {
        assertEquals(2147483647, IDL_LONG_MAX.value);
        assertEquals(-2147483648, IDL_LONG_MIN.value);
    }

    @Test
    void ulong_limits() {
        assertEquals(4294967295L, IDL_ULONG_MAX.value);
        assertEquals(0L, IDL_ULONG_MIN.value);
    }

    @Test
    void longlong_limits() {
        assertEquals(9223372036854775807L, LONGLONG_MAX.value);
        assertEquals(-9223372036854775808L, LONGLONG_MIN.value);
    }

    @Test
    @Disabled("figure out how to best represent unsigned long long in Java")
    void ulonglong_limits() {
        assertEquals(new java.math.BigInteger("18446744073709551615"), ULONGLONG_MAX.value);
        assertEquals(java.math.BigInteger.ZERO, ULONGLONG_MIN.value);
    }

    @Test
    void hex_literals() {
        assertEquals(0xDEADBEEFL, HEX_DEADBEEF.value);
        assertEquals(0xFFFFFFFFL, HEX_FFFFFFFF.value);
        assertNotNull(HEX_64_BIT.value);
    }

    @Test
    void octal_literals() {
        assertEquals(511, OCTAL_777.value);
    }

    @Test
    void large_int_struct() {
        var s = new LargeIntFields();
        assertEquals(0L, s.getBigSigned());
        assertEquals(0L, s.getBigUnsigned());
        s.setBigSigned(Long.MAX_VALUE);
        s.setBigUnsigned(Long.MIN_VALUE);
        assertEquals(Long.MAX_VALUE, s.getBigSigned());
        assertEquals(Long.MIN_VALUE, s.getBigUnsigned());
    }

    @Test
    void derived_constants() {
        assertEquals(2147483646, LONG_MAX_MINUS_ONE.value);
        assertEquals(9223372036854775806L, LONGLONG_MAX_MINUS_ONE.value);
    }
}
